File: app/src/main/java/com/zedge/contentstudio/domain/ArchiveClassifier.kt
package com.zedge.contentstudio.domain

import com.github.junrar.Archive
import com.zedge.contentstudio.core.ContentTypes
import com.zedge.contentstudio.core.NaturalOrder
import com.zedge.contentstudio.core.SetMeta
import com.zedge.contentstudio.data.LocalFile
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.util.zip.ZipInputStream

/** One media file inside a ZIP / RAR (path is the full path inside the archive). */
class ArchiveEntry(val name: String, val bytes: ByteArray, val archive: String) {
    val base: String get() = name.substringAfterLast('/')
    val dir: String get() = if (name.contains('/')) name.substringBeforeLast('/') else ""
    val size: Long get() = bytes.size.toLong()
    fun toLocalFile(): LocalFile = LocalFile(base, bytes, LocalFile.mimeFor(base))
}

/** Detected set (24H / Dual / Battery): exactly one entry per slot. */
class SetUnit(val type: String, val files: Map<String, ArchiveEntry>, val byName: Boolean, val label: String, val archive: String) {
    val meta: SetMeta get() = ContentTypes.SET_TYPES.getValue(type)
}

sealed class ImportUnit {
    class Set(val set: SetUnit) : ImportUnit()
    /** media = image | audio | video */
    class File(val media: String, val entry: ArchiveEntry) : ImportUnit()

    val title: String
        get() = when (this) {
            is Set -> set.label.ifBlank { set.archive.ifBlank { set.meta.label } }.let { if (set.label.isBlank()) "${set.meta.short} set (${it})" else "${set.meta.short}: $it" }
            is File -> entry.base
        }
}

class Classification(val units: List<ImportUnit>, val notes: List<String>, val mediaCount: Int)
class Grouping(val sets: List<SetUnit>, val leftovers: List<ArchiveEntry>, val imageCount: Int)

/**
 * Exact port of the dashboard's Smart Archive Import classification:
 *   folder with 2 images -> Dual, 4 -> 24H, 6 -> Battery; other images -> single wallpapers,
 *   mp3 -> ringtone, mp4/mov -> video. Works for ANY number of folders.
 */
object ArchiveClassifier {
    val IMG_RE = Regex("\\.(jpe?g|png|webp)$", RegexOption.IGNORE_CASE)
    val AUDIO_RE = Regex("\\.mp3$", RegexOption.IGNORE_CASE)
    val VIDEO_RE = Regex("\\.(mp4|mov)$", RegexOption.IGNORE_CASE)
    val AUX_RE = Regex("(preview|thumb|thumbnail|cover|poster|collage|mockup|screenshot)", RegexOption.IGNORE_CASE)

    fun isJunk(path: String): Boolean {
        val b = path.substringAfterLast('/')
        return b.isEmpty() || path.contains("__MACOSX/") || b.startsWith(".") || b.equals("thumbs.db", ignoreCase = true)
    }

    fun prettySetLabel(folder: String): String {
        if (folder.isBlank()) return ""
        return folder.replace(Regex("[_\\-]+"), " ").replace(Regex("\\s+"), " ").trim()
            .replace(Regex("\\b([a-z])")) { it.groupValues[1].uppercase() }
    }

    /** "set1_night.jpg" -> night, "lock.png" -> lock, "wall_2.jpg" -> 2nd slot. */
    fun detectSlot(fileBase: String, slots: List<String>): String? {
        val stem = fileBase.replace(Regex("\\.[^.]+$"), "").lowercase()
        val tokens = stem.split(Regex("[^a-z0-9]+")).filter { it.isNotEmpty() }
        for (slot in slots) {
            val words = ContentTypes.SLOT_WORDS[slot] ?: listOf(slot)
            if (words.any { it in tokens }) return slot
        }
        val m = Regex("(\\d+)$").find(stem)
        if (m != null) {
            val idx = m.groupValues[1].toInt() - 1
            if (idx >= 0 && idx < slots.size) return slots[idx]
        }
        return null
    }

    fun pickBySlotNames(files: List<ArchiveEntry>, slots: List<String>): Map<String, ArchiveEntry>? {
        val byName = LinkedHashMap<String, ArchiveEntry>()
        for (f in files) { val sl = detectSlot(f.base, slots); if (sl != null && !byName.containsKey(sl)) byName[sl] = f }
        return if (byName.size == slots.size) byName else null
    }

    fun makeSet(type: String, files: List<ArchiveEntry>, label: String, archive: String): SetUnit {
        val slots = ContentTypes.SET_TYPES.getValue(type).slots
        val byName = pickBySlotNames(files, slots)
        if (byName != null) return SetUnit(type, byName, true, label, archive)
        val ordered = LinkedHashMap<String, ArchiveEntry>()
        slots.forEachIndexed { i, sl -> ordered[sl] = files[i] }
        return SetUnit(type, ordered, false, label, archive)
    }

    private fun byDir(imgs: List<ArchiveEntry>): Map<String, List<ArchiveEntry>> {
        val m = LinkedHashMap<String, MutableList<ArchiveEntry>>()
        for (e in imgs) m.getOrPut(e.dir) { ArrayList() }.add(e)
        return m.keys.sortedWith(NaturalOrder).associateWith { k -> m.getValue(k).sortedWith(compareBy(NaturalOrder) { it.base }) }
    }

    /** Folder with exactly N images = one set; otherwise chunk N at a time in natural order. */
    fun groupIntoSets(entries: List<ArchiveEntry>, type: String, archive: String): Grouping {
        val slots = ContentTypes.SET_TYPES.getValue(type).slots
        val n = slots.size
        val imgs = entries.filter { !isJunk(it.name) && IMG_RE.containsMatchIn(it.name) }
        val sets = ArrayList<SetUnit>()
        val leftovers = ArrayList<ArchiveEntry>()
        for ((dir, files) in byDir(imgs)) {
            val folderLabel = prettySetLabel(dir.substringAfterLast('/'))
            if (files.size == n) { sets.add(makeSet(type, files, folderLabel, archive)); continue }
            if (files.size > n && files.size < 2 * n) {
                val main = files.filter { !AUX_RE.containsMatchIn(it.base) }
                val aux = files.filter { AUX_RE.containsMatchIn(it.base) }
                if (main.size == n) { sets.add(makeSet(type, main, folderLabel, archive)); leftovers.addAll(aux); continue }
                val picked = pickBySlotNames(files, slots)
                if (picked != null) {
                    val used = picked.values.toSet()
                    sets.add(SetUnit(type, picked, true, folderLabel, archive))
                    leftovers.addAll(files.filter { it !in used }); continue
                }
            }
            var k = 0
            var i = 0
            while (i + n <= files.size) {
                sets.add(makeSet(type, files.subList(i, i + n), if (folderLabel.isNotBlank()) "$folderLabel #${k + 1}" else "", archive))
                i += n; k++
            }
            leftovers.addAll(files.drop((files.size / n) * n))
        }
        return Grouping(sets, leftovers, imgs.size)
    }

    fun classify(entries: List<ArchiveEntry>, archiveName: String, fallbackType: String?): Classification {
        val units = ArrayList<ImportUnit>()
        val notes = ArrayList<String>()
        val media = entries.filter { !isJunk(it.name) && (IMG_RE.containsMatchIn(it.name) || AUDIO_RE.containsMatchIn(it.name) || VIDEO_RE.containsMatchIn(it.name)) }
        media.filter { AUDIO_RE.containsMatchIn(it.name) }.sortedWith(compareBy(NaturalOrder) { it.name }).forEach { units.add(ImportUnit.File("audio", it)) }
        media.filter { VIDEO_RE.containsMatchIn(it.name) }.sortedWith(compareBy(NaturalOrder) { it.name }).forEach { units.add(ImportUnit.File("video", it)) }
        val imgs = media.filter { IMG_RE.containsMatchIn(it.name) }
        val dirs = byDir(imgs)
        for ((dir, sorted) in dirs) {
            var files = sorted
            val label = prettySetLabel(dir.substringAfterLast('/'))
            var type = ContentTypes.COUNT_TO_SET_TYPE[files.size]
            if (type == null && dir.isNotEmpty()) {
                val main = files.filter { !AUX_RE.containsMatchIn(it.base) }
                if (main.size != files.size && ContentTypes.COUNT_TO_SET_TYPE.containsKey(main.size)) {
                    notes.add("$label: ignored ${files.size - main.size} preview/cover image(s)")
                    files = main
                    type = ContentTypes.COUNT_TO_SET_TYPE[files.size]
                }
            }
            if (type != null && dir.isNotEmpty()) { units.add(ImportUnit.Set(makeSet(type, files, label, archiveName))); continue }
            if (type != null && dir.isEmpty() && dirs.size == 1) { units.add(ImportUnit.Set(makeSet(type, files, "", archiveName))); continue }
            if (fallbackType != null && ContentTypes.isSet(fallbackType)) {
                val flat = files.map { ArchiveEntry(it.base, it.bytes, it.archive) }
                val g = groupIntoSets(flat, fallbackType, archiveName)
                g.sets.forEachIndexed { i, st -> units.add(ImportUnit.Set(SetUnit(fallbackType, st.files, st.byName, if (label.isNotBlank()) "$label #${i + 1}" else "", archiveName))) }
                if (g.leftovers.isNotEmpty()) notes.add("${label.ifBlank { "root" }}: ${g.leftovers.size} image(s) left over (not a full ${ContentTypes.SET_TYPES.getValue(fallbackType).short} set) - skipped")
                continue
            }
            if (dir.isNotEmpty() && files.size > 1) notes.add("$label: ${files.size} images is not a set size (2 / 4 / 6) - queued as single wallpapers")
            files.forEach { units.add(ImportUnit.File("image", it)) }
        }
        return Classification(units, notes, media.size)
    }

    /** "3 x 24H Wallpaper Set, 2 single wallpaper(s), 1 ringtone(s)" */
    fun describe(units: List<ImportUnit>): String {
        val c = HashMap<String, Int>()
        for (u in units) { val k = when (u) { is ImportUnit.Set -> u.set.type; is ImportUnit.File -> u.media }; c[k] = (c[k] ?: 0) + 1 }
        val parts = ArrayList<String>()
        for (t in listOf("WALLPAPER_24H", "WALLPAPER_DUAL", "WALLPAPER_BATTERY")) c[t]?.let { parts.add("$it x ${ContentTypes.SET_TYPES.getValue(t).label}") }
        c["image"]?.let { parts.add("$it single wallpaper(s)") }
        c["audio"]?.let { parts.add("$it ringtone(s)") }
        c["video"]?.let { parts.add("$it video(s)") }
        return parts.joinToString(", ")
    }
}

/** Reads every non-directory entry of a ZIP or RAR into memory (media-only, junk skipped). */
object ArchiveReader {
    suspend fun read(file: LocalFile): List<ArchiveEntry> = withContext(Dispatchers.IO) {
        val b = file.bytes
        val isRar = b.size > 4 && b[0] == 0x52.toByte() && b[1] == 0x61.toByte() && b[2] == 0x72.toByte() && b[3] == 0x21.toByte()
        val isZip = b.size > 2 && b[0] == 0x50.toByte() && b[1] == 0x4B.toByte()
        when {
            isRar || (!isZip && file.lower.endsWith(".rar")) -> readRar(b, file.name)
            isZip || file.lower.endsWith(".zip") -> readZip(b, file.name)
            else -> throw IllegalArgumentException("Only .zip and .rar archives are supported")
        }
    }

    private fun wanted(name: String): Boolean =
        !ArchiveClassifier.isJunk(name) && (ArchiveClassifier.IMG_RE.containsMatchIn(name) || ArchiveClassifier.AUDIO_RE.containsMatchIn(name) || ArchiveClassifier.VIDEO_RE.containsMatchIn(name) || name.lowercase().endsWith(".json"))  // v22: metadata sidecar

    private fun readZip(bytes: ByteArray, archive: String): List<ArchiveEntry> {
        val out = ArrayList<ArchiveEntry>()
        ZipInputStream(ByteArrayInputStream(bytes)).use { zis ->
            var e = zis.nextEntry
            while (e != null) {
                if (!e.isDirectory && wanted(e.name)) {
                    val bos = ByteArrayOutputStream(maxOf(1024, e.size.toInt().coerceAtLeast(0)))
                    zis.copyTo(bos)
                    out.add(ArchiveEntry(e.name, bos.toByteArray(), archive))
                }
                zis.closeEntry()
                e = zis.nextEntry
            }
        }
        if (out.isEmpty() && bytes.size < 22) throw IllegalArgumentException("Not a valid ZIP file")
        return out
    }

    private fun readRar(bytes: ByteArray, archive: String): List<ArchiveEntry> {
        val out = ArrayList<ArchiveEntry>()
        Archive(ByteArrayInputStream(bytes)).use { rar ->
            if (rar.isEncrypted) throw IllegalArgumentException("Encrypted RAR archives are not supported")
            var fh = rar.nextFileHeader()
            while (fh != null) {
                val name = (fh.fileName ?: "").replace('\\', '/')
                if (!fh.isDirectory && wanted(name)) {
                    val bos = ByteArrayOutputStream()
                    rar.extractFile(fh, bos)
                    out.add(ArchiveEntry(name, bos.toByteArray(), archive))
                }
                fh = rar.nextFileHeader()
            }
        }
        return out
    }
}

File: lib/domain/archive_classifier.dart
import '../core/constants.dart';
import '../data/archive_reader.dart';

/// Natural ("numeric aware", case-insensitive) compare - `Intl.Collator({numeric:true, sensitivity:"base"})`.
int natCmp(String a, String b) {
  final ra = RegExp(r'(\d+)|(\D+)');
  final ta = ra.allMatches(a.toLowerCase()).map((m) => m.group(0)!).toList();
  final tb = ra.allMatches(b.toLowerCase()).map((m) => m.group(0)!).toList();
  for (var i = 0; i < ta.length && i < tb.length; i++) {
    final x = ta[i], y = tb[i];
    final nx = int.tryParse(x), ny = int.tryParse(y);
    int c;
    if (nx != null && ny != null) {
      c = nx.compareTo(ny);
      if (c == 0) c = x.length.compareTo(y.length);
    } else {
      c = x.compareTo(y);
    }
    if (c != 0) return c;
  }
  return ta.length.compareTo(tb.length);
}

/// Slot detection from file names, e.g. "set1_night.jpg", "lock.png", "battery-low.jpg", "wall_2.jpg" (2nd slot).
const Map<String, List<String>> kSlotWords = {
  'morning': ['morning', 'morn', 'dawn', 'sunrise', 'am'],
  'afternoon': ['afternoon', 'noon', 'midday', 'day'],
  'evening': ['evening', 'sunset', 'dusk', 'eve'],
  'night': ['night', 'midnight', 'dark'],
  'lock': ['lock', 'lockscreen', 'ls'],
  'home': ['home', 'homescreen', 'hs'],
  'critical': ['critical', 'crit', 'empty'],
  'low': ['low'],
  'mid': ['mid', 'medium', 'half'],
  'high': ['high'],
  'full': ['full', '100'],
  'charging': ['charging', 'charge', 'chg'],
};

String? detectSlot(String fileBase, List<String> slots) {
  final stem = fileBase.replaceAll(RegExp(r'\.[^.]+$'), '').toLowerCase();
  final tokens = stem.split(RegExp(r'[^a-z0-9]+')).where((t) => t.isNotEmpty).toList();
  for (final slot in slots) {
    final words = kSlotWords[slot] ?? [slot];
    if (words.any(tokens.contains)) return slot;
  }
  final m = RegExp(r'(\d+)$').firstMatch(stem);
  if (m != null) {
    final idx = int.parse(m.group(1)!) - 1;
    if (idx >= 0 && idx < slots.length) return slots[idx];
  }
  return null;
}

final RegExp kAuxImageRe = RegExp(r'(preview|thumb|thumbnail|cover|poster|collage|mockup|screenshot)', caseSensitive: false);

String prettySetLabel(String? folder) {
  if (folder == null || folder.isEmpty) return '';
  final s = folder.replaceAll(RegExp(r'[_\-]+'), ' ').replaceAll(RegExp(r'\s+'), ' ').trim();
  return s.replaceAllMapped(RegExp(r'\b([a-z])'), (m) => m.group(1)!.toUpperCase());
}

/// A detected set: slot -> archive entry.
class SetGroup {
  SetGroup({required this.files, required this.label, required this.byName});
  final Map<String, ArchiveEntry> files;
  final String label;
  final bool byName;
}

Map<String, ArchiveEntry>? pickBySlotNames(List<ArchiveEntry> files, List<String> slots) {
  final byName = <String, ArchiveEntry>{};
  for (final f in files) {
    final sl = detectSlot(baseName(f.name), slots);
    if (sl != null && !byName.containsKey(sl)) byName[sl] = f;
  }
  return byName.length == slots.length ? byName : null;
}

SetGroup makeSetFromFiles(List<ArchiveEntry> files, List<String> slots, String label) {
  final byName = pickBySlotNames(files, slots);
  if (byName != null) return SetGroup(files: byName, label: label, byName: true);
  final ordered = <String, ArchiveEntry>{};
  for (var i = 0; i < slots.length; i++) {
    ordered[slots[i]] = files[i];
  }
  return SetGroup(files: ordered, label: label, byName: false);
}

class GroupResult {
  GroupResult(this.sets, this.leftovers, this.imageCount);
  final List<SetGroup> sets;
  final List<ArchiveEntry> leftovers;
  final int imageCount;
}

String _dirOf(String name) => name.contains('/') ? name.substring(0, name.lastIndexOf('/')) : '';

/// Folder with exactly N images = one set. Otherwise (flat list / big folder) chunk by N in natural name order.
GroupResult groupEntriesIntoSets(List<ArchiveEntry> entries, List<String> slots) {
  final n = slots.length;
  final imgs = entries.where((e) => !isJunkPath(e.name) && kImgNameRe.hasMatch(e.name)).toList();
  final byDir = <String, List<ArchiveEntry>>{};
  for (final e in imgs) {
    byDir.putIfAbsent(_dirOf(e.name), () => []).add(e);
  }
  final sets = <SetGroup>[];
  final leftovers = <ArchiveEntry>[];
  final dirs = byDir.keys.toList()..sort(natCmp);
  for (final dir in dirs) {
    final files = byDir[dir]!..sort((a, b) => natCmp(baseName(a.name), baseName(b.name)));
    final folderLabel = prettySetLabel(dir.isNotEmpty ? dir.split('/').last : '');
    if (files.length == n) {
      sets.add(makeSetFromFiles(files, slots, folderLabel));
      continue;
    }
    if (files.length > n && files.length < 2 * n) {
      final main = files.where((f) => !kAuxImageRe.hasMatch(baseName(f.name))).toList();
      final aux = files.where((f) => kAuxImageRe.hasMatch(baseName(f.name))).toList();
      if (main.length == n) {
        sets.add(makeSetFromFiles(main, slots, folderLabel));
        leftovers.addAll(aux);
        continue;
      }
      final picked = pickBySlotNames(files, slots);
      if (picked != null) {
        final used = picked.values.toSet();
        sets.add(SetGroup(files: picked, label: folderLabel, byName: true));
        leftovers.addAll(files.where((f) => !used.contains(f)));
        continue;
      }
    }
    var k = 0;
    for (var i = 0; i + n <= files.length; i += n, k++) {
      sets.add(makeSetFromFiles(files.sublist(i, i + n), slots, folderLabel.isNotEmpty ? '$folderLabel #${k + 1}' : ''));
    }
    leftovers.addAll(files.sublist((files.length ~/ n) * n));
  }
  return GroupResult(sets, leftovers, imgs.length);
}

/// Smart archive import unit: either a whole set or a single media file.
class ImportUnit {
  ImportUnit.file({required this.media, required this.entry})
      : kind = 'file',
        type = null,
        files = null,
        byName = false,
        label = '',
        archive = entry?.zArchive;

  ImportUnit.set({required this.type, required this.files, required this.byName, required this.label, required this.archive})
      : kind = 'set',
        media = null,
        entry = null;

  final String kind; // file | set
  final String? media; // audio | video | image
  final ArchiveEntry? entry;
  final String? type; // WALLPAPER_DUAL | WALLPAPER_24H | WALLPAPER_BATTERY
  final Map<String, ArchiveEntry>? files;
  final bool byName;
  final String label;
  final String? archive;

  bool get isSet => kind == 'set';

  /// `unitTitle`
  String get title {
    if (isSet) return '${kSetTypeMeta[type]!.label}${label.isNotEmpty ? ' - $label' : ''}';
    return baseName(entry!.name);
  }
}

class ClassifyResult {
  ClassifyResult(this.units, this.notes, this.mediaCount);
  final List<ImportUnit> units;
  final List<String> notes;
  final int mediaCount;
}

const Map<int, String> kCountToSetType = {2: 'WALLPAPER_DUAL', 4: 'WALLPAPER_24H', 6: 'WALLPAPER_BATTERY'};

/// Classify everything inside a ZIP/RAR without any user selection:
/// folder with 2 images -> Dual set, 4 -> 24H set, 6 -> Battery set;
/// other images -> single wallpapers, mp3 -> ringtone, mp4/mov -> video.
ClassifyResult classifyArchiveEntries(List<ArchiveEntry> entries, String archiveName, String? fallbackType) {
  final units = <ImportUnit>[];
  final notes = <String>[];
  final media = entries
      .where((e) => !isJunkPath(e.name) && (kImgNameRe.hasMatch(e.name) || kAudioNameRe.hasMatch(e.name) || kVideoNameRe.hasMatch(e.name)))
      .toList();
  for (final e in media) {
    e.zArchive = archiveName;
  }
  final audios = media.where((e) => kAudioNameRe.hasMatch(e.name)).toList()..sort((a, b) => natCmp(a.name, b.name));
  for (final e in audios) {
    units.add(ImportUnit.file(media: 'audio', entry: e));
  }
  final videos = media.where((e) => kVideoNameRe.hasMatch(e.name)).toList()..sort((a, b) => natCmp(a.name, b.name));
  for (final e in videos) {
    units.add(ImportUnit.file(media: 'video', entry: e));
  }
  final imgs = media.where((e) => kImgNameRe.hasMatch(e.name)).toList();
  final byDir = <String, List<ArchiveEntry>>{};
  for (final e in imgs) {
    byDir.putIfAbsent(_dirOf(e.name), () => []).add(e);
  }
  final dirs = byDir.keys.toList()..sort(natCmp);
  for (final dir in dirs) {
    var files = byDir[dir]!..sort((a, b) => natCmp(baseName(a.name), baseName(b.name)));
    final label = prettySetLabel(dir.isNotEmpty ? dir.split('/').last : '');
    String? type = kCountToSetType[files.length];
    if (type == null && dir.isNotEmpty) {
      final main = files.where((f) => !kAuxImageRe.hasMatch(baseName(f.name))).toList();
      if (main.length != files.length && kCountToSetType.containsKey(main.length)) {
        notes.add('$label: ignored ${files.length - main.length} preview/cover image(s)');
        files = main;
        type = kCountToSetType[files.length];
      }
    }
    if (type != null && dir.isNotEmpty) {
      final st = makeSetFromFiles(files, kSetTypeMeta[type]!.slots, label);
      units.add(ImportUnit.set(type: type, files: st.files, byName: st.byName, label: label, archive: archiveName));
      continue;
    }
    if (type != null && dir.isEmpty && byDir.length == 1) {
      final st = makeSetFromFiles(files, kSetTypeMeta[type]!.slots, '');
      units.add(ImportUnit.set(type: type, files: st.files, byName: st.byName, label: '', archive: archiveName));
      continue;
    }
    if (fallbackType != null && kSetTypeMeta.containsKey(fallbackType)) {
      final fm = kSetTypeMeta[fallbackType]!;
      final g = groupEntriesIntoSets(files.map((f) => f.copyWithName(baseName(f.name))).toList(), fm.slots);
      for (var i = 0; i < g.sets.length; i++) {
        final st = g.sets[i];
        units.add(ImportUnit.set(
            type: fallbackType, files: st.files, byName: st.byName, label: label.isNotEmpty ? '$label #${i + 1}' : '', archive: archiveName));
      }
      if (g.leftovers.isNotEmpty) {
        notes.add('${label.isNotEmpty ? label : 'root'}: ${g.leftovers.length} image(s) left over (not a full ${fm.short} set) - skipped');
      }
      continue;
    }
    if (dir.isNotEmpty && files.length > 1) {
      notes.add('$label: ${files.length} images is not a set size (2 / 4 / 6) - queued as single wallpapers');
    }
    for (final e in files) {
      units.add(ImportUnit.file(media: 'image', entry: e));
    }
  }
  return ClassifyResult(units, notes, media.length);
}

String describeUnits(List<ImportUnit> units) {
  final c = <String, int>{};
  for (final u in units) {
    final k = u.isSet ? u.type! : u.media!;
    c[k] = (c[k] ?? 0) + 1;
  }
  final parts = <String>[];
  for (final t in const ['WALLPAPER_24H', 'WALLPAPER_DUAL', 'WALLPAPER_BATTERY']) {
    if ((c[t] ?? 0) > 0) parts.add('${c[t]} x ${kSetTypeMeta[t]!.label}');
  }
  if ((c['image'] ?? 0) > 0) parts.add('${c['image']} single wallpaper(s)');
  if ((c['audio'] ?? 0) > 0) parts.add('${c['audio']} ringtone(s)');
  if ((c['video'] ?? 0) > 0) parts.add('${c['video']} video(s)');
  return parts.join(', ');
}
